beacon-chain/db/filesystem/data_column.go

```diff
@@ -0,0 +1,966 @@
+	count, err = file.WriteAt(sszEncodedDataColumnSidecars, metadata.fileSize)
+	if err != nil {
+		return errors.Wrap(err, "write SSZ encoded data column sidecars")
+	}
+	if count != len(sszEncodedDataColumnSidecars) {
+		return ErrWrongBytesWritten
+	}
+
+	return nil
+}
+
+// saveDataColumnSidecarsNewFile saves data column sidecars into a new file.
+// This function expects all data column sidecars to belong to the same block.
+func (dcs *DataColumnStorage) saveDataColumnSidecarsNewFile(filePath string, inputDataColumnSidecars chan []blocks.VerifiedRODataColumn) (err error) {
+	// Initialize the indices.
+	var indices [mandatoryNumberOfColumns]byte
+
+	var (
+		sszEncodedDataColumnSidecarRefSize int
+		sszEncodedDataColumnSidecars       []byte
+	)
+
+	// Initialize the count of the saved SSZ encoded data column sidecar.
+	storedCount := 0
+
+	log.Error("AAAA")
+
+	for {
+		dataColumnSidecars := pullChan(inputDataColumnSidecars)
+		if len(dataColumnSidecars) == 0 {
+			break
+		}
+
+		for _, dataColumnSidecar := range dataColumnSidecars {
+			// Extract the data column index.
+			dataColumnIndex := dataColumnSidecar.ColumnIndex
+
+			// Skip if the data column is already stored.
+			if indices[dataColumnIndex] >= limit {
+				continue
+			}
+
+			// Alter the indices to mark the first data column sidecar as saved.
+			// savedCount can safely be cast to uint8 since it is less than limit.
+			indices[dataColumnIndex] = limit + uint8(storedCount)
+
+			// Increment the count of the saved SSZ encoded data column sidecar.
+			storedCount++
+
+			// SSZ encode the first data column sidecar.
+			sszEncodedDataColumnSidecar, err := dataColumnSidecar.MarshalSSZ()
+			if err != nil {
+				return errors.Wrap(err, "data column sidecar marshal SSZ")
+			}
+
+			// Check if the size of the SSZ encoded data column sidecar is correct.
+			if sszEncodedDataColumnSidecarRefSize != 0 && len(sszEncodedDataColumnSidecar) != sszEncodedDataColumnSidecarRefSize {
+				return ErrWrongSszEncodedDataColumnSidecarSize
+			}
+
+			// Set the SSZ encoded data column sidecar reference size.
+			sszEncodedDataColumnSidecarRefSize = len(sszEncodedDataColumnSidecar)
+
+			// Append the first SSZ encoded data column sidecar to the SSZ encoded data column sidecars.
+			sszEncodedDataColumnSidecars = append(sszEncodedDataColumnSidecars, sszEncodedDataColumnSidecar...)
+		}
+	}
+
+	if storedCount == 0 {
+		// Nothing to save.
+		return nil
+	}
+
+	// Create the data column sidecars file.
+	dir := filepath.Dir(filePath)
+	if err := dcs.fs.MkdirAll(dir, directoryPermissions()); err != nil {
+		return errors.Wrapf(err, "mkdir all")
+	}
+
+	file, err := dcs.fs.Create(filePath)
+	if err != nil {
+		return errors.Wrap(err, "data column sidecars file path create")
+	}
+
+	defer func() {
+		closeErr := file.Close()
+
+		// Overwrite the existing error only if it is nil, since the close error is less important.
+		if closeErr != nil && err == nil {
+			err = closeErr
```

Comment:
Hesitate to even suggest this as error on close is rare, but maybe log it?
```
if closeErr == nil {
        return
}
if err != nil {
        log.WithError(err).WithField("path", filePath).Error("error closing file")
        return
}
err = closeErr
```

Reply:
I was inspired by a master for this one:
https://github.com/prysmaticlabs/prysm/blob/6d16ed4a89504b2c7c067f5a15ea7bd7b65b049b/beacon-chain/db/filesystem/blob.go#L179-L185

I don't have any strong opinion on that. But I like the fact the close error is really handled as a real error.